compiler+runtime/src/cpp/jank/aot/processor.cpp

```diff
@@ -0,0 +1,305 @@
+    }
+
+    for(auto const &lib : { "-ljank",
+                            /* Default libraries that jank depends on. */
+                            "-lfolly",
+                            "-lstdc++",
+                            "-lm",
+                            "-lzip",
+                            "-lzippp_static",
+                            "-lcrypto",
+                            "-lclang-cpp",
+                            "-lgc",
+                            "-lgccpp",
+                            "-lnanobench",
+                            "-lftxui-component",
+                            "-lLLVM",
+                            "-lftxui-dom",
+                            "-lftxui-screen",
+                            "-lcpptrace",
+                            "-ldwarf" })
+    {
+      Args.push_back(strdup(lib));
+    }
+
+    for(auto const &define : define_macros)
+    {
+      Args.push_back(strdup(util::format("-D{}", define).c_str()));
+    }
+
+    Args.push_back(strdup("-std=c++20"));
+
+    Args.push_back(strdup("-o"));
+    auto const output_filepath{ relative_to_cache_dir(output_filename) };
+    Args.push_back(strdup(output_filepath.c_str()));
+
+    /* Required because of `strdup` usage and need to manually free the memory.
+     * Clang expects C strings that we own. */
+    util::scope_exit const cleanup{ [&]() {
+      for(auto const s : Args)
+      {
+        /* NOLINTNEXTLINE(cppcoreguidelines-no-malloc) */
+        free(reinterpret_cast<void *>(const_cast<char *>(s)));
+      }
+    } };
+
+    llvm::ArrayRef<char const *> const Argv(Args);
+
+    std::unique_ptr<clang::driver::Compilation> C(driver.BuildCompilation(Argv));
+
+    if(!C || C->containsError())
+    {
+      return compiler_err{ 1, "Failed to build compilation steps." };
+    }
+
+    /* Execute the compilation jobs (preprocess, compile, assemble)
+     * This actually runs the commands determined by BuildCompilation. */
+    int Result = 0;
+    if(C && !C->containsError())
+    {
+      llvm::SmallVector<std::pair<int, clang::driver::Command const *>> FailingCommands;
+      Result = driver.ExecuteCompilation(*C, FailingCommands);
+
+      for(auto const &P : FailingCommands)
+      {
+        /* Check if command signaled an error */
+        if(P.first < 0)
+        {
+          llvm::errs() << "Error executing command: " << P.second->getExecutable() << "\n";
+
+          /* TODO: We can print more details about the failing command P.second */
+        }
+      }
+    }
+    else
+    {
+      Result = 1;
+    }
+
+    if(diags.hasErrorOccurred())
+    {
+      return compiler_err{ 1, "Compilation failed with errors.\n" };
+    }
+
+    if(Result == 0)
+    {
+      llvm::outs() << util::format("Compilation successful. Find the executable at: `{}`",
+                                   output_filepath)
+                        .c_str();
+    }
+    else
+    {
+      return compiler_err{ Result, "Compilation command execution failed.\n" };
+    }
```

Comment:
Rather than having an if for positive, an else for negative, and then a positive return, let's minimize the branching. Check for your early returns. Everything else doesn't get nested.

So, if we check that the result != 0, we can early return with the error. Otherwise, we can print before we return `ok()` (the `jtl::` should not be needed).